Remove symmath from edx-platform

Carlos-Muniz opened the pull request:
## Description

This PR removes `symmath` from `common/lib/`, all removable mentions of `symmath` within `edx-platform` and upgrades the requirements to no longer use a non-existent symmath.

This is the second part of [BOM-2581](https://openedx.atlassian.net/browse/BOM-2581). 
Symmath has been added to openedx-calc alongside calc in [openedx-calc PR #38](https://github.com/openedx/openedx-calc/pull/38).

## Testing Instructions
The code has been tested:
- Manually: 
  - A custom dev image of lms/cms was built using this branch via tutor
  - Tutor was started in dev mode using these images
  - While these containers were running, the version of `openedx-calc` was checked to be `3.0.1`, and that `symmath` was not in `common/lib/`
  - In the CMS GUI (Studio), a new course was formed, and a new symbolic math problem was initiated.
  - Answers that were mathematically correct, but symbolically different were entered into the sample problem
- Locally: Pytest via Tutor
- Online: All checks have passed

## Additional information
There is an issue with pip-tools using pip version >=22.0.0. Pip is currently constrained to <20.3.4, and so this shouldn't be an issue, but this may still be an issue.

There is currently an issue with the package m2r, which was causing make upgrade to fail. The dependency mistune has been constrained to <2.0.0 in order for everything to work as it once did.

In the near future, these may be fixed.

Commit:
refactor: Remove symmath from common/lib

Commit:
refactor: Remove all mentions of symmath

Commit:
chore: Update requirements to remove symmath

Commit:
fix: Correct path to symmath function

jmbowman commented:
Cool!  Could you drop a note in https://openedx.atlassian.net/browse/BOM-2581 once this is merged so we can close it out?  There are more tasks like this in that ticket's parent epic if you want to keep going. ^_^

Oh, and there's a new pip-tools release that should resolve the issue you noted: https://github.com/jazzband/pip-tools/releases/tag/6.5.0

kdmccormick commented on common/lib/capa/capa/responsetypes.py:
```diff
@@ -2546,7 +2546,7 @@ def setup_response(self):
 
     def execute_check_function(self, idset, submission):
-        from symmath import symmath_check
+        from symmath.symmath_check import symmath_check
```
Hm, what is making this import change necessary?

Keep in mind that because [symmath has been available in the edx-sandbox (ie, codejail) environment](https://github.com/openedx/edx-platform/blob/5708787527453bbaa11ed90d1d98a35ec3ca1388/requirements/edx-sandbox/py38.txt#L9), we must assume that symmath's interface is used in some instructor-authored grading code that we do not have visibility into. Because of that, we cannot make any breaking symmath API changes.

Carlos-Muniz replied:
When `symmath` was moved to `openedx-calc`, the following was removed from `symmath/__init__.py`
```
# lint-amnesty, pylint: disable=missing-module-docstring
from .formula import *
from .symmath_check import *
```

This allows the importation directly from `symmath_check`.
This can be re-added quickly to `openedx-calc`.

kdmccormick replied:
Sounds great.

Carlos-Muniz replied:
Related: https://github.com/openedx/openedx-calc/pull/41

kdmccormick commented on requirements/constraints.txt:
```diff
@@ -97,5 +97,9 @@ py2neo<2022
+# m2r fails to specify the version of mistune that it needs leading to the error message:
+# AttributeError: module 'mistune' has no attribute 'BlockGrammar'
+# https://github.com/miyakogi/m2r/issues/66
+mistune<2.0.0
```
Good details here. Another detail that's good to include is the answer to the question "when can we remove this constraint?" (if someone is reading comments in constraints.txt, there's a good change they're looking to de-constrain something ;)

The comment above `docutils<0.18` is a good example of this.

Carlos-Muniz replied:
I'm not sure there is a clear answer to [this](https://github.com/miyakogi/m2r/issues/66).
Either [m2r locks mistune or m2r updates to mistune 2.0 API](https://github.com/miyakogi/m2r/pull/43).
Neither seems likely at the moment without some outside help.

kdmccormick commented on requirements/edx/local.in:
```diff
@@ -3,7 +3,6 @@
 -e common/lib/capa
 -e common/lib/safe_lxml
 -e common/lib/sandbox-packages
--e common/lib/symmath
```
🎉

Carlos-Muniz replied:
🎉

kdmccormick commented on scripts/verify-dunder-init.sh:
```diff
@@ -38,7 +38,7 @@ exclude+='|^common/test/data/?.*$'
 #  * common/lib/xmodule -> EXCLUDE from check.
 #  * common/lib/xmodule/xmodule/modulestore -> INCLUDE in check.
 exclude+='|^common/lib$'
-exclude+='|^common/lib/(capa|safe_lxml|sandbox-packages|symmath|xmodule)$'
```
Grepping for 'symmath', I found one other instance like this you can clean up in `scripts/post-pip-compile.sh`.

Carlos-Muniz replied:
The mention, and the associated logic have been removed.

Commit:
docs: Update mistune constraint description

Commit:
feat: Update openedx-calc==3.0.1

This allows common/lib/capa/capa/responsetypes.py to not have to
change.

kdmccormick commented on scripts/post-pip-compile.sh:
```diff
@@ -14,17 +14,6 @@ function show_help {
-        mv ${TEMP_FILE} ${FILE_PATH}
-        sed "s|$BASE_FILE_URL|.|" ${FILE_PATH} > ${TEMP_FILE}
-        mv ${TEMP_FILE} ${FILE_PATH}
-    fi
```
So I was curious why this line specifically calls out `common/lib/symmath`, as there are four other local projects in edx-platform:
* `common/lib/xmodule`
* `common/lib/capa`
* `common/lib/safe_lxml`
* `common/lib/sandbox-packages`
* `openedx/core/lib/xblock_builtin`

all of which suffer from the same type of weird interaction with pip tools that is linked in the comment.

If I'm reading this block of code correctly, it seems like we are only using `common/lib/symmath` to see:
* does `FILE_CONTENTS` contain local Python requirements that need fixing because they contain absolute paths? and
* if so, where on the filesystem is the root of edx-platform? (`BASE_REMATCH[1]` takes on the value of the first regex match in parentheses) 
 

Then, with `BASE_FILE_URL` determined, we fix *all* local requirement paths. I think, rather than deleting this block, it needs to be updated to work without `common/lib/symmath`.

Carlos-Muniz replied:
I understand now. I was under the impression that this was specific to `symmath`.
Would it be sufficient to replace `symmath` with one of the path's that is still there?
Like `capa`?

kdmccormick replied:
@Carlos-Muniz Yup, I think that'd be the right move. If I had to recommend one I'd say `xmodule` because that one is the biggest and thus the one most likely to move last.

Carlos-Muniz replied:
good thinking. I'll keep the code-block, and change it to xmodule.

kdmccormick reviewed (changes requested):
Did you test this out manually? The `## Test instructions` section of the PR description is a great place to communicate whether & how you manually verified that your code works -- sometimes I use it less as "test instructions" and more as "test explanation".

Commit:
feat: Change symmath to xmodule

scripts/post-pip-compile.sh uses common/lib/symmath/ as its sample
directory to clean up after using pip-tools. Since symmath has been
removed, xmodule, the largest of the directories in common/lib/ will
replace it.

kdmccormick commented:
Code looks good! Only blocker for me is this question: https://github.com/openedx/edx-platform/pull/29869#pullrequestreview-875129278

Carlos-Muniz commented:
> Code looks good! Only blocker for me is this question: [#29869 (review)](https://github.com/openedx/edx-platform/pull/29869#pullrequestreview-875129278)

The code has been tested:
- Manually: openedx-calc 3.0.1 has been installed in a local env and made sure that it was able to be imported as it is in the code.
- Locally: Pytest via Tutor
- Online: All checks have passed.

kdmccormick commented:
@Carlos-Muniz Could you go into LMS/Studio and confirm that a symbolic math problem renders and can be responded to?

kdmccormick reviewed (approved):
Pending tests, looks good to me 👍🏻

Carlos-Muniz commented:
Closes https://github.com/openedx/tcril-engineering/issues/65

jdmulloy commented:
@Carlos-Muniz @kdmccormick I can't figure out why, but something in this broke the edxapp build on GoCD. This is the ansible output where it's failing. We may need to revert this.

`TASK [edxapp : code sandbox | Install base sandbox requirements and create sandbox virtualenv] ***
fatal: [10.4.11.90]: FAILED! => {
    "changed": false,
    "cmd": [
        "/edx/app/edxapp/venvs/edxapp-sandbox/bin/pip3",
        "install",
        "-i",
        "https://pypi.python.org/simple",
        "--exists-action",
        "w",
        "-r",
        "/edx/app/edxapp/edx-platform/requirements/edx-sandbox/py38.txt"
    ]
}

MSG:

stdout: Looking in indexes: https://pypi.python.org/simple
Obtaining file:///edx/app/edxapp/edx-platform/common/lib/sandbox-packages (from -r /edx/app/edxapp/edx-platform/requirements/edx-sandbox/py38.txt (line 7))
Collecting cffi==1.15.0
  Downloading cffi-1.15.0-cp38-cp38-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (446 kB)
Collecting chem==1.2.0
  Downloading chem-1.2.0-py3-none-any.whl (24 kB)
Collecting click==7.1.2
  Downloading click-7.1.2-py2.py3-none-any.whl (82 kB)
Collecting cryptography==36.0.1
  Downloading cryptography-36.0.1-cp36-abi3-manylinux_2_24_x86_64.whl (3.6 MB)
Collecting cycler==0.11.0
  Downloading cycler-0.11.0-py3-none-any.whl (6.4 kB)
Collecting decorator==4.4.2
  Downloading decorator-4.4.2-py2.py3-none-any.whl (9.2 kB)
Collecting joblib==1.1.0
  Downloading joblib-1.1.0-py2.py3-none-any.whl (306 kB)
Collecting kiwisolver==1.3.2
  Downloading kiwisolver-1.3.2-cp38-cp38-manylinux_2_5_x86_64.manylinux1_x86_64.whl (1.2 MB)
Collecting lxml==4.5.0
  Downloading lxml-4.5.0-cp38-cp38-manylinux1_x86_64.whl (5.6 MB)
Collecting markupsafe==2.0.1
  Downloading MarkupSafe-2.0.1-cp38-cp38-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (30 kB)
Collecting matplotlib==3.3.4
  Downloading matplotlib-3.3.4-cp38-cp38-manylinux1_x86_64.whl (11.6 MB)
Collecting mpmath==1.2.1
  Downloading mpmath-1.2.1-py3-none-any.whl (532 kB)
Collecting networkx==2.5.1
  Downloading networkx-2.5.1-py3-none-any.whl (1.6 MB)
Collecting nltk==3.7
  Downloading nltk-3.7-py3-none-any.whl (1.5 MB)
Collecting numpy==1.16.6
  Downloading numpy-1.16.6.zip (5.1 MB)
Collecting openedx-calc==3.0.1
  Downloading openedx_calc-3.0.1-py3-none-any.whl (34 kB)
Collecting pillow==9.0.1
  Downloading Pillow-9.0.1-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.3 MB)
Collecting pycparser==2.21
  Downloading pycparser-2.21-py2.py3-none-any.whl (118 kB)
Collecting pyparsing==3.0.7
  Downloading pyparsing-3.0.7-py3-none-any.whl (98 kB)
Collecting python-dateutil==2.4.0
  Downloading python_dateutil-2.4.0-py2.py3-none-any.whl (175 kB)
Collecting random2==1.0.1
  Downloading random2-1.0.1.zip (21 kB)
Collecting regex==2022.1.18
  Downloading regex-2022.1.18-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (764 kB)
Collecting scipy==1.7.3
  Downloading scipy-1.7.3-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (39.3 MB)
Collecting six==1.16.0
  Downloading six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting sympy==1.6.2
  Downloading sympy-1.6.2-py3-none-any.whl (5.8 MB)
Collecting tqdm==4.62.3
  Downloading tqdm-4.62.3-py2.py3-none-any.whl (76 kB)
Building wheels for collected packages: numpy, random2
  Building wheel for numpy (setup.py): started
  Building wheel for numpy (setup.py): still running...
  Building wheel for numpy (setup.py): finished with status 'done'
  Created wheel for numpy: filename=numpy-1.16.6-cp38-cp38-linux_x86_64.whl size=9934057 sha256=78575335ee1e7b9ac3b9a0556268394a6a87d170c94da58f6a142ba0ac4aad84
  Stored in directory: /edx/app/edxapp/venvs/edxapp-sandbox/.cache/pip/wheels/22/6f/a6/069db0c95f9bb6b73e07da014891c58460fef1eb84f49576f0
  Building wheel for random2 (setup.py): started
  Building wheel for random2 (setup.py): finished with status 'done'
  Created wheel for random2: filename=random2-1.0.1-py3-none-any.whl size=12070 sha256=d6cc470f3c506b36072961e8357d8eec3521f4ba058eaf145f7d7a727a764047
  Stored in directory: /edx/app/edxapp/venvs/edxapp-sandbox/.cache/pip/wheels/6e/ca/f2/082dec051ffcaec249ae491b8c90e305726a7390274682b4a4
Successfully built numpy random2
Installing collected packages: tqdm, six, regex, pycparser, numpy, mpmath, joblib, click, sympy, scipy, python-dateutil, pyparsing, pillow, nltk, markupsafe, lxml, kiwisolver, decorator, cycler, cffi, sandbox-packages, random2, openedx-calc, networkx, matplotlib, cryptography, chem
  Running setup.py develop for sandbox-packages

:stderr:     ERROR: Command errored out with exit status 1:
     command: /edx/app/edxapp/venvs/edxapp-sandbox/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/edx/app/edxapp/edx-platform/common/lib/sandbox-packages/setup.py'"'"'; __file__='"'"'/edx/app/edxapp/edx-platform/common/lib/sandbox-packages/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps
         cwd: /edx/app/edxapp/edx-platform/common/lib/sandbox-packages/
    Complete output (3 lines):
    running develop
    running egg_info
    error: [Errno 13] Permission denied
    ----------------------------------------
ERROR: Command errored out with exit status 1: /edx/app/edxapp/venvs/edxapp-sandbox/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/edx/app/edxapp/edx-platform/common/lib/sandbox-packages/setup.py'"'"'; __file__='"'"'/edx/app/edxapp/edx-platform/common/lib/sandbox-packages/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps Check the logs for full command output.
WARNING: You are using pip version 20.3.4; however, version 22.0.3 is available.
You should consider upgrading via the '/edx/app/edxapp/venvs/edxapp-sandbox/bin/python -m pip install --upgrade pip' command.`

kdmccormick commented:
@jdmulloy weird! This is totally safe to revert if needed -- no data migrations or anything.

@Carlos-Muniz and I can take a look at a fix tomorrow.

pshiu commented:
Thanks, @kdmccormick, @Carlos-Muniz! Reverting now to see if we can get a green build.

kdmccormick commented:
This line in the logs jumped out at me:
```
:stderr: ERROR: Command errored out with exit status 1:
command: /edx/app/edxapp/venvs/edxapp-sandbox/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/edx/app/edxapp/edx-platform/common/lib/sandbox-packages/setup.py'"'"'; file='"'"'/edx/app/edxapp/edx-platform/common/lib/sandbox-packages/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(file);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, file, '"'"'exec'"'"'))' develop --no-deps
cwd: /edx/app/edxapp/edx-platform/common/lib/sandbox-packages/
Complete output (3 lines):
running develop
running egg_info
error: [Errno 13] Permission denied
```

Seems like something changed with `./common/lib/sandbox-packages`.

Lo and behold: https://github.com/openedx/edx-platform/pull/29869/files#diff-b637ba9a442d8914ab88f5231e76fa60adc6d862fc13fdcaff4bb27329edd4beL7-R7. Notice the additional `-e ` that this PR adds.

kdmccormick commented:
@Carlos-Muniz  ^

Carlos-Muniz commented:
Great catch @kdmccormick!
